Review images documentation. (#487)
MicroEJ/docs: VEEPortingGuide/uiImageRaw.rst

It manages two formats in output: the MicroEJ format and the binary format.
The Image Renderer manages only the MicroEJ format (:ref:`section_image_standard_raw`, :ref:`section_image_display_raw` and :ref:`section_image_gpu_raw`).
The binary output format (:ref:`section_image_binary_raw`) is fully platform dependent and can be used to encode some images which are not usable by MicroUI standard API.

The output format is generated from the input format:

* by using the off-board tool :ref:`section_image_generator` at application compile-time 
* or by using a :ref:`runtime decoder <image_runtime_decoder>` of the :ref:`section_image_loader` at application run-time.

The embedded format may be the one of the output formats or the :ref:`original input format <section_image_asis>`.

.. _section_image_display_raw:

MicroEJ Format: Display
=======================

See :ref:`section_image_display_output`.

The display can hold a pixel encoding which is not standard (see :ref:`display_pixel_structure`).
The MicroEJ format can be customized to encode the pixel in same encoding than display.

> The number of bits-per-pixels and the pixel bits organisation is asked during the MicroEJ format generation and when the ``drawImage`` algorithms are running.

Reviewer: ```suggestion
The number of bits per pixel and the pixel bit organization is asked during the MicroEJ format generation and when the ``drawImage`` algorithms are running.
```

If the image to encode contains some transparent pixels, the output file will embed the transparency according to the display’s implementation capacity.
When all pixels are fully opaque, no extra information will be stored in the output file in order to free up some memory space.

> .. note:: From Image Engine point of view, the format stays a MicroEJ format, readable by the Image Renderer.

Reviewer: ```suggestion
.. note:: From the Image Engine point of view, the format stays a MicroEJ format, readable by the Image Renderer.
```

The required memory is similar to :ref:`section_image_standard_raw`.

.. _section_image_standard_raw:

MicroEJ Format: Standard
========================

See :ref:`section_image_standard_output`.

> This format requires a small header (around 20 bytes) to store the image size (width, height), format, flags (is_transparent etc.), row stride etc.

Reviewer: ```suggestion
This format requires a small header (around 20 bytes) to store the image size (width, height), format, flags (is_transparent etc.), row stride, etc.
```

> The required memory also depends on the number of bits-per-pixels of MicroEJ format:

Reviewer: ```suggestion
The required memory also depends on the number of bits per pixels of the MicroEJ format:
```

                  ;
      }

-  RGB888: 24-bit format, 8 per color.
   ::

      u32 convertARGB8888toRAWFormat(u32 c){
          return c & 0xffffff;
      }

> -  RGB565: 16-bit format, 5 or 6 per color.

Reviewer: ```suggestion
-  RGB565: 16-bit format, 5 for red, 6 for green, 5 for blue.
```

                  ;
      }

-  A8: 8-bit format, only transparency is encoded.
   ::

      u32 convertARGB8888toRAWFormat(u32 c){
          return 0xff - (toGrayscale(c) & 0xff);
      }

-  A4: 4-bit format, only transparency is encoded.
   ::

      u32 convertARGB8888toRAWFormat(u32 c){
          return (0xff - (toGrayscale(c) & 0xff)) / 0x11;
      }

-  A2: 2-bit format, only transparency is encoded.
   ::

      u32 convertARGB8888toRAWFormat(u32 c){
          return (0xff - (toGrayscale(c) & 0xff)) / 0x55;
      }

-  A1: 1-bit format, only transparency is encoded.
   ::

      u32 convertARGB8888toRAWFormat(u32 c){
          return (0xff - (toGrayscale(c) & 0xff)) / 0xff;
      }

> The pixels order follows this rule:

Reviewer: ```suggestion
The pixel order follows this rule:
```

   ::

         pixel_offset = (pixel_Y * image_width + pixel_X) * bpp / 8;

MicroEJ Format: Grayscale
=========================

See :ref:`section_image_grayscale_output`.

> This format requires a small header (around 20 bytes) to store the image size (width, height), format, flags (is_transparent etc.), row stride etc.

Reviewer: ```suggestion
This format requires a small header (around 20 bytes) to store the image size (width, height), format, flags (is_transparent etc.), row stride, etc.
```

> The required memory also depends on the number of bits-per-pixels of MicroEJ format:

Reviewer: ```suggestion
The required memory also depends on the number of bits per pixels of the MicroEJ format:
```

              ;
      }

-  C4: 4 bits with grayscale conversion.
   ::

      u32 convertARGB8888toRAWFormat(u32 c){
          return (toGrayscale(c) & 0xff) / 0x11;
      }

-  C2: 2 bits with grayscale conversion.
   ::

      u32 convertARGB8888toRAWFormat(u32 c){
          return (toGrayscale(c) & 0xff) / 0x55;
      }

-  C1: 1 bit with grayscale conversion.
   ::

      u32 convertARGB8888toRAWFormat(u32 c){
          return (toGrayscale(c) & 0xff) / 0xff;
      }

> The pixels order follows this rule:

Reviewer: ```suggestion
The pixel order follows this rule:
```

   ::

         pixel_offset = (pixel_Y * image_width + pixel_X) * bpp / 8;

MicroEJ Format: RLE Compressed
==============================

See :ref:`section_image_rle_output`.

.. _section_image_gpu_raw:

MicroEJ Format: GPU
===================

> The MicroEJ format may be customized to be platform's GPU compatible. It can be extended by one or several restrictions on the pixels array: 

Reviewer: ```suggestion
The MicroEJ format may be customized to be compatible with the platform's GPU. It can be extended by one or several restrictions on the pixels array: 
```

* Its start address has to be aligned on a higher value than the number of bits-per-pixels. 
* A padding has to be added after each line (row stride).

> * The MicroEJ format can hold a platform dependant header, located between MicroEJ format header (start of file) and pixels array.

Reviewer: ```suggestion
* The MicroEJ format can hold a platform-dependent header, located between the MicroEJ format header (start of file) and the pixel array.
```

>   The MicroEJ format is designed to let the platform encodes and decodes this additional header.

Reviewer: ```suggestion
  The MicroEJ format is designed to let the platform encode and decode this additional header.
```

>   For Image Engine software algorithms, this header is useless and never used. 

Reviewer: ```suggestion
  For Image Engine software algorithms, this header is not needed and never used.
```

Disadvantages:

* No compression: the image size in bytes is proportional to the number of pixels.
  The required memory is similar to :ref:`section_image_standard_raw` when there is no custom header. 

> When MicroEJ format holds another header (called ``custom_header``), the required memory is:

Reviewer: ```suggestion
When the MicroEJ format holds another header (called ``custom_header``), the required memory is:
```

::

      required_memory = header + custom_header + (image_width * image_height) * bpp / 8;

> The row stride allows to add some padding at the end of each line in order to start next line at an address with a specific memory alignment; it is often required by hardware accelerators (GPU).

Reviewer: ```suggestion
The row stride allows to add some padding at the end of each line in order to start the next line at an address with a specific memory alignment; it is often required by hardware accelerators (GPU).
```

The row stride is by default a value in relation with the image width: ``row_stride_in_bytes = image_width * bpp / 8``.
It can be customized at image buffer creation thanks to the Low Level API ``LLUI_DISPLAY_IMPL_getNewImageStrideInBytes``.

> The required RAM memory becomes:

Reviewer: ```suggestion
The required memory becomes:
```

::

      required_memory = header + custom_header + row_stride * image_height;

.. figure:: images/uiFormat02.*
   :width: 50.0%

.. _section_image_binary_raw:

Binary Format
=============

> This format is not compatible with the Image Renderer and by MicroUI.

Reviewer: ```suggestion
This format is not compatible with the Image Renderer and MicroUI.
```

> It is can be used by MicroUI addon libraries which provide their own images managements. 

Reviewer: ```suggestion
It is can be used by MicroUI addon libraries which provide their own image management procedures.
```

Author: `It is can be` also ;)
